src/protobuf.h

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright The async-profiler authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef _PROTOBUF_H
+#define _PROTOBUF_H
+
+#include "arch.h"
+#include <string.h>
+#include <type_traits>
+
+typedef const u32 protobuf_t;
+static protobuf_t VARINT = 0;
+static protobuf_t LEN = 2;
+
+typedef u32 protobuf_index_t;
+
+// We assume the length of a nested field can be represented with 3 varint
+// bytes.
+const size_t nested_field_byte_count = 3;
+
+class ProtobufBuffer {
+private:
+  ProtobufBuffer *_parent_message;
+  char *_data;
+  size_t _offset;
+
+  ProtobufBuffer(ProtobufBuffer *parent) :
+    _data(parent->_data + parent->offset()),
+    _offset(0),
+    _parent_message(parent) {}
+
+  template <typename T>
+  typename std::enable_if<std::is_unsigned<T>::value, void>::type
```

Comment:
Is this template really useful? I think `u64` variant covers all cases. Writing `putVarInt<>` with angle brackets does not look nice.

Reply:
`u64` variant covers all cases, but can we accept the additional overhead of having a new `u64` stack variable when we have a `u32` already? In the meantime I removed the templating: ad85b8800cac40ccef1e0079834d6ec54948c003